M365/MDO/MDOThreatPolicyChecker.ps1

```diff
@@ -0,0 +1,840 @@
+
+    [Parameter(ParameterSetName = 'AppliedCsv')]
+    [Parameter(ParameterSetName = 'AppliedEmail')]
+    [Parameter(ParameterSetName = 'AppliedMDOCsv')]
+    [Parameter(ParameterSetName = 'AppliedMDOEmail')]
+    [Parameter(ParameterSetName = 'AppliedTenant')]
+    [switch]$SkipVersionCheck,
+
+    [Parameter(Mandatory = $true, ParameterSetName = "ScriptUpdateOnly")]
+    [switch]$ScriptUpdateOnly
+)
+
+begin {
+
+    . $PSScriptRoot\..\..\Shared\ScriptUpdateFunctions\Test-ScriptVersion.ps1
+    . $PSScriptRoot\..\..\Shared\LoggerFunctions.ps1
+    . $PSScriptRoot\..\..\Shared\OutputOverrides\Write-Host.ps1
+
+    # Cache to reduce calls to Get-MgGroup
+    $groupCache = @{}
+
+    function Get-GroupObjectId {
+        [OutputType([string])]
+        param(
+            [Parameter(Mandatory = $true)]
+            [ValidateNotNullOrEmpty()]
+            [MailAddress]$groupEmail
+        )
+
+        # Check the cache first
+        if ($groupCache.ContainsKey($groupEmail)) {
+            return $groupCache[$groupEmail]
+        }
+
+        # Get the group
+        $group = $null
+        $group = Get-MgGroup -Filter "mail eq '$($groupEmail)'" -ErrorAction SilentlyContinue
+
+        if ($group) {
+            # Cache the result
+            $groupCache[$groupEmail] = $group.Id
+
+            # Return the Object ID of the group
+            return $group.Id
+        } else {
+            Write-Host "The EmailAddress of group $groupEmail was not found" -ForegroundColor Red
+            return $null
+        }
+    }
+
+    function Test-EmailAddress {
+        [OutputType([MailAddress])]
+        param(
+            [Parameter(Mandatory = $true)]
+            [ValidateNotNullOrEmpty()]
+            [string]$EmailAddress,
+            [Parameter(Mandatory = $true)]
+            [ValidateNotNullOrEmpty()]
+            $AcceptedDomains
+        )
+
+        try {
+            $tempAddress = $null
+            $tempAddress = [MailAddress]$EmailAddress
+        } catch {
+            Write-Host "The EmailAddress $EmailAddress cannot be validated. Please provide a valid email address." -ForegroundColor Red
+            return $null
+        }
+        $recipient = $null
+        $recipient = Get-Recipient $EmailAddress -ErrorAction SilentlyContinue
+        if ($null -eq $recipient) {
+            Write-Host "$EmailAddress is not a recipient in this tenant" -ForegroundColor Red
+            return $null
+        } else {
+            $Domain = $tempAddress.Host
+            if ($AcceptedDomains.DomainName -contains $Domain) {
+                return $tempAddress
+            } else {
+                Write-Host "The domain $Domain is not an accepted domain in your organization. Please provide a valid email address." -ForegroundColor Red
+                return $null
+            }
+        }
+    }
+
+    # Function to check if an email is in a group
+    # Cache of members to reduce number of calls to Get-MgGroupMember
+    $memberCache = @{}
```

Comment:
Can we move this initialization of the hashtable to the top or bottom of the `begin` section? It's easy to overlook this because it sits between all of these functions.